core/integration/host_test.go

```diff
@@ -299,6 +299,89 @@ func (HostSuite) TestDirectoryExcludeInclude(ctx context.Context, t *testctx.T)
 	})
 }
 
+func (HostSuite) TestDirectoryNoGitAutoIgnore(ctx context.Context, t *testctx.T) {
+	dir := t.TempDir()
+
+	require.NoError(t, os.MkdirAll(filepath.Join(dir, ".git"), 0o755))
+	require.NoError(t, os.WriteFile(filepath.Join(dir, ".gitignore"), []byte(`**.txt
+.git/
+!subdir/e.txt
+*.md
+`), 0o600))
+	require.NoError(t, os.WriteFile(filepath.Join(dir, "b.md"), []byte("2"), 0o600))
+	require.NoError(t, os.WriteFile(filepath.Join(dir, "c.txt.rar"), []byte("3"), 0o600))
+	require.NoError(t, os.MkdirAll(filepath.Join(dir, "subdir"), 0o755))
+	require.NoError(t, os.WriteFile(filepath.Join(dir, "subdir", "b.md"), []byte("1"), 0o600))
+	require.NoError(t, os.WriteFile(filepath.Join(dir, "subdir", "e.txt"), []byte("2"), 0o600))
+	require.NoError(t, os.WriteFile(filepath.Join(dir, "subdir", "g.txt"), []byte("6"), 0o600))
+	require.NoError(t, os.WriteFile(filepath.Join(dir, "subdir", "h.yaml"), []byte("6"), 0o600))
+	require.NoError(t, os.MkdirAll(filepath.Join(dir, "subdir2"), 0o755))
+	require.NoError(t, os.WriteFile(filepath.Join(dir, "subdir2", ".gitignore"), []byte(`*.yaml
+.gitignore
+`), 0o600))
+	require.NoError(t, os.WriteFile(filepath.Join(dir, "subdir2", "foo.go"), []byte("1"), 0o600))
+	require.NoError(t, os.WriteFile(filepath.Join(dir, "subdir2", "bar.txt"), []byte("1"), 0o600))
+	require.NoError(t, os.WriteFile(filepath.Join(dir, "subdir2", "baz.md"), []byte("1"), 0o600))
+	require.NoError(t, os.WriteFile(filepath.Join(dir, "subdir2", "bool.yaml"), []byte("1"), 0o600))
+
+	c := connect(ctx, t)
+
+	t.Run("disable git auto ignore", func(ctx context.Context, t *testctx.T) {
+		entries, err := c.Host().Directory(dir, dagger.HostDirectoryOpts{NoGitAutoIgnore: true}).Entries(ctx)
+		require.NoError(t, err)
+		require.Equal(t, []string{".git/", ".gitignore", "b.md", "c.txt.rar", "subdir/", "subdir2/"}, entries)
+
+		subDirEntries, err := c.Host().Directory(filepath.Join(dir, "subdir"),
+			dagger.HostDirectoryOpts{NoGitAutoIgnore: true},
+		).Entries(ctx)
+		require.NoError(t, err)
+		require.Equal(t, []string{"b.md", "e.txt", "g.txt", "h.yaml"}, subDirEntries)
+
+		subDir2Entries, err := c.Host().Directory(filepath.Join(dir, "subdir2"),
+			dagger.HostDirectoryOpts{NoGitAutoIgnore: true},
+		).Entries(ctx)
+		require.NoError(t, err)
+		require.Equal(t, []string{".gitignore", "bar.txt", "baz.md", "bool.yaml", "foo.go"}, subDir2Entries)
+	})
+
+	t.Run("apply auto git ignore by default", func(ctx context.Context, t *testctx.T) {
+		hostDir := c.Host().Directory(dir)
+
+		rootHostDir, err := hostDir.Entries(ctx)
+		require.NoError(t, err)
+		require.Equal(t, []string{".gitignore", "c.txt.rar", "subdir/", "subdir2/"}, rootHostDir)
+
+		subDirEntries, err := hostDir.Directory("subdir").Entries(ctx)
+		require.NoError(t, err)
+		// require.Equal(t, []string{"b.md", "e.txt", "h.yaml"}, subDirEntries) // XXX: wrong?
+		require.Equal(t, []string{"e.txt", "h.yaml"}, subDirEntries)
```

Comment:
These tests are directly imported from #10736. But I think they're wrong, at least from my testing using `git` directly?

We previously ignore `*.md` - so `b.md` shouldn't be found.

*Unlike* in `dockerignore`, `*.md` in `gitignore` should ignore all markdown files, no matter where they are:

```
❯ cd `mktemp -d`
❯ git init
Initialized empty Git repository in /tmp/tmp.JGS60Al5B1/.git/
❯ echo '*.md' >> .gitignore
❯ touch a.md
❯ mkdir b
❯ touch b/c.md
❯ git add .
❯ git status
On branch main

No commits yet

Changes to be committed:
  (use "git rm --cached <file>..." to unstage)
	new file:   .gitignore
```